src/string_object.cpp

```diff
@@ -579,6 +579,29 @@ size_t StringObject::char_count(Env *env) const {
     return char_count;
 }
 
+Value StringObject::setbyte(Env *env, Value index_obj, Value value_obj) {
+    assert_not_frozen(env);
+
+    nat_int_t index = IntegerObject::convert_to_nat_int_t(env, index_obj);
+    nat_int_t value = IntegerObject::convert_to_nat_int_t(env, value_obj);
+    nat_int_t length = static_cast<nat_int_t>(m_string.length());
+
+    if (index + length < 0) {
+        env->raise("IndexError", "index {} out of string", index);
+    }
+
+    if (index >= length) {
+        env->raise("IndexError", "index {} out of string", index);
+    }
+
+    if (index < 0) {
+        index += length;
+    }
```

Comment:
Ahh, here is where we check for a negative index. But let's say the `index` is `-100` and the `length` is only `10`... the resulting `index` will still be negative. :grimacing: 

I think you need to move this code up higher, then check the result is positive.

Reply:
Yeah, negative index is a bit tricky.

Index is allowed to have the following values:
```
-length <= index < length
```

Otherwise `IndexError` should be raised.

This incrementing `index += length;` may be placed at the beginning but we still need an original value to form the proper exception message. So just try to avoid adding new local variable with original value here.

Some examples:

```ruby
s = "abc"; s.setbyte(-1, "d".ord); s
# => "abd"
s = "abc"; s.setbyte(-2, "d".ord); s
# => "adc"
s = "abc"; s.setbyte(-3, "d".ord); s
# => "dbc"
s = "abc"; s.setbyte(-4, "d".ord); s
# (irb):11:in `setbyte': index -4 out of string (IndexError)
```

Reply:
OK, I think I get it! This is fine then.

I tend to lean toward readability, so I think I would have created a temporary variable to hold the original value. Or, I would have used `index_obj` for the error message.

But in any case, this works and now I know how to read `if (index + length < 0)` which was a bit too clever for my brain. :laughing: 

:shipit:

Reply:
👍